build-logic/basics/src/main/kotlin/gradlebuild/basics/BuildEnvironment.kt

```diff
@@ -20,11 +20,34 @@ import gradlebuild.basics.BuildParams.CI_ENVIRONMENT_VARIABLE
 import org.gradle.api.JavaVersion
 import org.gradle.api.Project
 import org.gradle.api.file.Directory
-import org.gradle.api.provider.Provider
+import org.gradle.api.file.DirectoryProperty
+import org.gradle.api.provider.Property
 import org.gradle.internal.os.OperatingSystem
 
 
-fun Project.repoRoot() = layout.projectDirectory.parentOrRoot()
+abstract class BuildEnvironmentExtension {
+    abstract val repoRoot: DirectoryProperty
+    abstract val gitCommitId: Property<String>
+    abstract val gitBranch: Property<String>
+}
+
+
+/**
+ * Get a property cached in `BuildEnvironmentExtension`, or query the property via heavy I/O operations then put it into cache.
+ */
+private
+fun <T> Project.getCachedPropertyOrQuery(propertyGetter: BuildEnvironmentExtension.() -> Property<T>, valueSupplier: Project.() -> T): T {
```

Comment:
I don't see why we need this. If you want something to only run once, you can add a plugin to the root project. For the caching, I think you can use `finalizeValueOnRead()`, to ensure that the query only runs once.

Reply:
I use Kotlin `lazy {}` to cache the value.